woocommerce/payment-gateway/payment-tokens/class-sv-wc-payment-gateway-payment-token.php

```diff
@@ -540,6 +540,22 @@ public function set_billing_hash( $value ) {
 	}
 
 
+	/**
+	 * Gets the gateway environment that this token is associated with.
+	 *
+	 * @since 5.6.0-dev.1
+	 *
+	 * @return string|null
+	 */
+	public function get_environment() {
```

Comment:
hmm I thought we were going with the same route as the other setters to read from `$data` once `read()` has done what this method in this PR is doing here

True, we currently don't pass `$environment` since this is part of the tokens handler; it'd be valuable that we do so moving forward, wouldn't it?

So, hypothetically, this method could return:

```php

$environment = isset( $this->data['environment'] ) && is_string( $this->data['environment'] ) ? $this->data['environment'] : '';

if ( '' === $environment && ( $handler = $this->get_tokens_handler() ) ) {
	$environment = $handler->get_gateway()->get_environment();
}

return $environment;

```

Perhaps the `::build_token()` method in the handler could pass the current environment when using `$data` as an array, since it has context of the environment?

And also, perhaps the `get_environment()` method could fallback on the value currently set in the handler? Because it's essentially what's happening already in the context of this class when we use the method internally: if hypothetically returns empty, then the handler will still get the default defined on the handler itself. Since this method is public, it could have other uses though.

maybe @ChaseWiseman could offer a 2nd opinion here

@wvega was also suggesting we could have a setter too

Reply:
>I thought we were going with the same route as the other setters to read from $data once read() has done what this method in this PR is doing here

I agree that we need to be able to read the environment when the framework token is instantiated.

I decided to follow the recommendations in the story because the tests and other modifications would still be useful even after changing `get_environment()` and adding a setter or taking the value in the constructor.

>And also, perhaps the get_environment() method could fallback on the value currently set in the handler?

I don't think we should have `get_environment()` fallback to `$this->get_tokens_handler()->get_gateway()->get_environment()`. One reason is because it currently forces us to use partial mocks for testing, which I think is better if we can avoid them. Also, I think we currently only pass the environment to the handler itself, which defaults to the gateway environment anyway, so long as we pass `null` whenever we don't know the environment.

The version of the method returns `null` as fallback, but using an empty string to maintain the return type works for me too. We can use `$this->get_environment() ?: null` to call `update_legacy_token()` and `delete_legacy_token()` and get the same behavior (or update those methods to handle empty strings the same way).